Add docs for upgrading to Caracal (#1250)
stackhpc/stackhpc-kayobe-config: doc/source/operations/upgrading-openstack.rst

There are many changes in the OpenStack |current_release| release described in
the release notes for each project. Here are some notable ones.

> RabbitMQ SLURP upgrade

Reviewer: This feels like it's in the wrong section (Notable changes in the |current_release| Release)

Reviewer: Perhaps it should go into prerequisites?

----------------------

Because this is a SLURP upgrade, RabbitMQ must be upgraded manually from 3.11,
to 3.12, then to 3.13 on Antelope before the Caracal upgrade. This upgrade
should not cause an API outage (though it should still be considered "at
risk").

There are two prerequisites:

> 1. Kolla-Ansible should be upgraded to the latest version. Assuming a typical
>    production layout, execute the following from the kolla-ansible source
>    directory:

Reviewer 2: ```suggestion
1. Kolla-Ansible should be upgraded to the latest version:
```

      cd $KOLLA_SOURCE_PATH
      git fetch && git pull

>       $KOLLA_VENV_PATH/bin/pip install .

Reviewer: `kayobe control host bootstrap`?

>       git fetch && git pull && ../../venvs/kolla-ansible/bin/pip install .

Reviewer 2: ```suggestion
      cd $KOLLA_SOURCE_PATH
      git fetch && git pull
      $KOLLA_VENV_PATH/bin/pip install .
```

Reviewer 2: Should remove "assuming a typical production layout" if you accept this change

2. The RabbitMQ container image tag must be equal to or newer than
   ``20240823T101942``. Check the timestamps in

>    ``etc/kayobe/kolla-image-tags.yml``.

Reviewer: Would it make sense to do this after updating to latest antelope? (L212) Then this would always be true.

Once complete, upgrade RabbitMQ:

.. code-block:: bash

>    kayobe overcloud container image pull -kt rabbitmq

Reviewer 2: ```suggestion
   kayobe overcloud service configuration generate --node-config-dir /tmp/ignore -kt none
```
This container image pull will only bring in the 3.11 image that we won't even use. Speaking to @Alex-Welsh in the office, he put this here to make sure that the new local kolla config is generated.
I suggest we switch to a more explicit command for this, it's the same workaround we've used elsewhere when needing to generate kolla config while using ``kayobe kolla ansible run``.

Can't wait for Caracal, where we get this patched: https://review.opendev.org/c/openstack/kayobe/+/896394

Author: Done

   kayobe kolla ansibe run "rabbitmq-upgrade 3.12"

>    kayobe kolla ansibe run "rabbitmq-upgrade 3.13"

Reviewer: `s/ansibe/ansible/g`

RabbitMQ quorum queues
----------------------

In Caracal, quorum queues are enabled by default for RabbitMQ. This is
different to Antelope which used HA queues. Before upgrading to Caracal, it is
strongly recommended that you migrate from HA to quorum queues. The migration
is automated using a script.

.. warning::
   This migration will stop all services using RabbitMQ and cause an
   extended API outage while queues are migrated. It should only be
   performed in a pre-agreed maintenance window.

Set the following variables in your kolla globals file (i.e.
``$KAYOBE_CONFIG_PATH/kolla/globals.yml`` or
``$KAYOBE_CONFIG_PATH/environments/$KAYOBE_ENVIRONMENT/kolla/globals.yml``):

.. code-block:: yaml

      om_enable_rabbitmq_high_availability: false
      om_enable_rabbitmq_quorum_queues: true

Then execute the migration script:

.. code-block:: bash

>    $KAYOBE_CONFIG_PATH/../../tools/rabbitmq-quorum-migration.sh

Reviewer 2: Given we document this here, and now have a handy dandy migration script, we can delete these docs: https://github.com/stackhpc/stackhpc-kayobe-config/blob/stackhpc/2024.1/doc/source/operations/rabbitmq.rst

Author: Deleted

TODO: guide for disabling Heat

Designate sink disabled by default
----------------------------------

Designate sink is optional designate service which listens for event
Notifications, primarily from Nova and Neutron. It is disabled by default (when

> designate is enabled) in Caracal. It is not required for Designate to function.

Reviewer: Designate sink is an optional Designate service which listens for event
notifications, primarily from Nova and Neutron. It is disabled by default (when
designate is enabled) in Caracal. It is not required for Designate to function.

This behavior can be overridden manually:

.. code-block:: yaml
   :caption: ``kolla/globals.yml``

   designate_enable_notifications_sink: true

> If you are using Designate and do not make this change, the Antelope
> ``designate-sink`` container will remain on the controllers after the upgrade.
> It must be removed manually.

Reviewer 2: Looks like this was disabled by default simply because it's not required for Designate to work. I've proposed a wording change to make this clearer.

```suggestion
This was disabled by default simply because it isn't required for Designate to function. Unless you are using this feature, you do not need to make this change. Note that the Antelope
``designate-sink`` container will remain on the controllers after the upgrade.
It must be removed manually.
```

Author: I've gone with slightly different wording, what do you think?

Reviewer 2: Alex's update is good to me

> It must be removed manually.

Reviewer: Kolla-ansible so lazy :D
